### PR TITLE
[RW-1190] Concept set API [risk=low]

### PR DESCRIPTION
**Pull request by danqrodney**
(Working on tests now.)

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      throw new BadRequestException(String.format(
+          "Concept set \"/%s/%s/%d\" already exists.",
+          workspaceNamespace, workspaceId, dbConceptSet.getName()));
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  private ConceptSet toClientConceptSet(org.pmiops.workbench.db.model.ConceptSet conceptSet) {
+    ConceptSet result = TO_CLIENT_CONCEPT_SET.apply(conceptSet);
+    if (!conceptSet.getConceptIds().isEmpty()) {
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Is there a reason not to do this in `TO_CLIENT_CONCEPT_SET`?

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  public ResponseEntity<ConceptSetListResponse> getConceptSetsInWorkspace(String workspaceNamespace,
+      String workspaceId) {
+    workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(workspaceNamespace, workspaceId,
+        WorkspaceAccessLevel.READER);
+
+    Workspace workspace = workspaceService.getRequiredWithConceptSets(workspaceNamespace, workspaceId);
+    ConceptSetListResponse response = new ConceptSetListResponse();
+    Set<org.pmiops.workbench.db.model.ConceptSet> conceptSets = workspace.getConceptSets();
+    if (conceptSets != null) {
+      Set<Long> conceptIds = Sets.newHashSet();
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Is the relationship here that each conceptSet has many conceptIds?

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 3]
```diff
@@ -581,6 +581,7 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
 
     org.pmiops.workbench.db.model.Workspace fromWorkspace = workspaceService.getRequiredWithCohorts(
         workspaceNamespace, workspaceId);
+    // CLONE concept sets
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Nit: Put TODO in front of this?

File: api/src/main/java/org/pmiops/workbench/db/model/StorageEnums.java [hunk 4]
```diff
@@ -51,6 +52,26 @@ public static Short authorityToStorage(Authority authority) {
     return CLIENT_TO_STORAGE_AUTHORITY.get(authority);
   }
 
+  private static final BiMap<ConceptSetDomain, Short> CLIENT_TO_STORAGE_CONCEPT_SET_DOMAIN =
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/db/model/StorageEnums.java]: Appreciate the work getting ahead of the enum issue!

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java (hunk 1 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: TO_CLIENT_CONCEPT_SET doesn't do any RPCs.... which is appropriate. We don't want to issue N RPCs when dealing with N concept sets.

In the list case, we use it to convert the concept sets without the concepts underneath, and then separately fetch the concepts for all N concept sets in one go from the DB. That way we're still doing a constant number of database lookups.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java (hunk 2 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: That is correct. We may want to place a limit on the number of concepts that can be in a single set; I will add that before I merge I think.

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java (hunk 3 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Will address this before merging.

**Comment by danqrodney** [comment]
Yes, tests are coming -- just wanted to get this out there for feedback on the API.

**Comment by blrubenstein** [comment]
Ah. Actually reading the PR comment now... My bad. Yes, I think the API looks good!

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java (hunk 1 above)
Comment (blrubenstein) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Got it. That sounds good! 👍

File: api/src/main/resources/workbench.yaml [hunk 5]
```diff
@@ -972,6 +1079,21 @@ definitions:
         REVIEW_ID_VERIFICATION
     ]
 
+  ConceptSetDomain:
```
Comment (calbach) [inline api/src/main/resources/workbench.yaml]: How does this interact with the dynamic domain CSV file which feeds into the CDR indexing pipeline? The list here doesn't match those.

https://github.com/all-of-us/workbench/blob/master/api/db-cdr/csv/domain.csv

File: api/src/main/java/org/pmiops/workbench/api/CohortsController.java [hunk 6]
```diff
@@ -292,7 +292,7 @@ public void setUserProvider(Provider<User> userProvider) {
 
     org.pmiops.workbench.db.model.Cohort cohort =
         cohortDao.findOne(cohortId);
-    if (cohort == null) {
+    if (cohort == null || cohort.getWorkspaceId() != workspace.getWorkspaceId()) {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/CohortsController.java]: I believe so -- if you requested a cohort under the wrong workspace, you'd still get it back. And yeah this is bad since I think it would let you access cohorts you don't actually have permission to use. 

I will add a test.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    dbConceptSet.setWorkspaceId(workspace.getWorkspaceId());
+    dbConceptSet.setCreationTime(now);
+    dbConceptSet.setLastModifiedTime(now);
+    dbConceptSet.setVersion(1);
+    try {
+      dbConceptSet = conceptSetDao.save(dbConceptSet);
+      // TODO: add recent resource entry for concept sets?
+    } catch (DataIntegrityViolationException e) {
+      throw new BadRequestException(String.format(
+          "Concept set \"/%s/%s/%d\" already exists.",
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Done.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+          workspaceNamespace, workspaceId, dbConceptSet.getName()));
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  private ConceptSet toClientConceptSet(org.pmiops.workbench.db.model.ConceptSet conceptSet) {
+    ConceptSet result = TO_CLIENT_CONCEPT_SET.apply(conceptSet);
+    if (!conceptSet.getConceptIds().isEmpty()) {
+      result.setConcepts(ImmutableList.copyOf(conceptDao.findAll(conceptSet.getConceptIds())).stream().map(
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Done.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    try {
+      dbConceptSet = conceptSetDao.save(dbConceptSet);
+    } catch (OptimisticLockException e) {
+      throw new ConflictException("Failed due to concurrent concept set modification");
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  @Override
+  public ResponseEntity<ConceptSet> updateConceptSetConcepts(String workspaceNamespace,
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: So I'm assuming that the UI for adding concepts to concept sets will look like find a concept or concepts, and add those specifically to a concept set; and that removing concepts from a concept set will look like selecting a concept or concepts already in a set, and hitting a Remove button. In either case, we're just sending up deltas, and concept sets could be pretty large (say hundreds of concepts). So I think it's better to send up just concepts to remove or add.

That doesn't work with the normal update() method, which expects to have at least the full set of concept IDs in the request. Hence this extra method. WDYT?

File: api/src/main/java/org/pmiops/workbench/db/model/StorageEnums.java [hunk 10]
```diff
@@ -51,6 +52,26 @@ public static Short authorityToStorage(Authority authority) {
...
+      .put(ConceptSetDomain.OBSERVATION, (short) 5)
+      .put(ConceptSetDomain.PROCEDURE, (short) 6)
+      .put(ConceptSetDomain.VISIT, (short) 7)
+      .build();
+
+  public static ConceptSetDomain conceptSetDomainFromStorage(Short domain) {
+    return CLIENT_TO_STORAGE_CONCEPT_SET_DOMAIN.inverse().get(domain);
+  }
+
+  public static Short conceptSetDomainToStorage(ConceptSetDomain domain) {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/model/StorageEnums.java]: Will do.

File: api/src/main/resources/workbench.yaml [hunk 11]
```diff
@@ -57,6 +57,13 @@ parameters:
     type: integer
     format: int64
     required: true
+  conceptSetId:
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: I actually want to follow the same pattern as cohorts; I think there are good reasons for it.

Concept set names (like cohort names) are not URL safe, and are mutable, so having them in the path probably doesn't make sense. (I want to be able to say 'rename the concept set with ID X to "Blah blah blah"'... this allows us to do that.)

You *could* refer to cohorts or concept set IDs by ID instead of by name in the cohort materialization API, but we don't expect researchers to actually know the internal ID of their cohort or concept set... so they would have to look it up programatically by name... at which you point you'd want to give them a convenience method that didn't require them to do that... and then you could save yourself a trip to the server by just accepting the cohort name or concept set name instead of the ID. So that's why we use IDs there.

Furthermore, when you clone a workspace, the cohort and concept set IDs will change, but the names of those things won't. If you want existing notebooks to keep working, they need to look them up by name, not ID.

So I think we have good reasons to use an immutable ID for the PK and REST path, but names when referencing cohorts and concept sets in notebooks. Let me know if this sounds reasonable to you.

File: api/src/main/resources/workbench.yaml (hunk 5 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: This table matches the cohort tables in our CDM schema that have standard / source concepts:

https://github.com/all-of-us/workbench/blob/master/api/config/cdm/cdm_5_2.json

Basically we're giving you all the tables you can pull stuff from when using a concept set.

The domains in the CSV file / OMOP data map to this, many-to-many... currently represented in this mapping:

https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/api/ConceptsController.java#L25

and when searching for concepts we will need to allow for multiple "domain" values in the OMOP data for a single "domain" indicated here.

We could pick a different term instead of "domain" for this, but I'm not thinking of a good one. 

Basically OMOP has concept -> domain -> one or more OMOP tables; they could have decided to instead do a many-to-many from concepts to OMOP tables, which is kind of what we're doing.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    Set<org.pmiops.workbench.db.model.ConceptSet> conceptSets = workspace.getConceptSets();
+    if (conceptSets != null) {
+      Set<Long> conceptIds = Sets.newHashSet();
+      // Get all the concept IDs for all concept sets, and look them all up at once.
+      for (org.pmiops.workbench.db.model.ConceptSet conceptSet : conceptSets) {
+        conceptIds.addAll(conceptSet.getConceptIds());
+      }
+      final Map<Long, Concept> conceptMap;
+      if (conceptIds.size() > 0) {
+        conceptMap = ImmutableList.copyOf(conceptDao.findAll(conceptIds)).stream()
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: No -- this will actually return whatever concepts it finds, and ignore everything else.

If we want to ensure that there are no concept set IDs in here which aren't in the concept table of the corresponding CDR DB, we can add some business logic to enforce that; but either way I don't think we'd blow up here.

I am not sure how we're going to handle the case of pointing workspaces at a new CDR version, which in theory might not have concepts that were in the previous one... in those cases there's something to be said for copying the IDs over and just not showing them in the UI.

File: api/src/main/resources/workbench.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: BTW, another option we considered early on (but decided against) was having an immutable human readable ID based off the name, scoped to the workspace. That could be used in both the REST path and notebooks (and would avoid issues where a cohort rename would break notebooks). But then you have a problem where when you change the name, the immutable ID no longer matches, and you can't change it. We decided that it was simpler and better to be able to change names and just warn users that doing so may break notebooks.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+  public ResponseEntity<ConceptSetListResponse> getConceptSetsInWorkspace(String workspaceNamespace,
+      String workspaceId) {
+    workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(workspaceNamespace, workspaceId,
+        WorkspaceAccessLevel.READER);
+
+    Workspace workspace = workspaceService.getRequiredWithConceptSets(workspaceNamespace, workspaceId);
+    ConceptSetListResponse response = new ConceptSetListResponse();
+    Set<org.pmiops.workbench.db.model.ConceptSet> conceptSets = workspace.getConceptSets();
+    if (conceptSets != null) {
+      // Concept sets in the list response will *not* have concepts under them, as this could be
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Note: I changed this behavior. I don't think we have a reason at this point to pre-fetch all the concepts in all the concept sets in a workspace, and in some cases that may be a lot of data.

We could consider changing this later.

File: api/src/main/resources/workbench.yaml (hunk 5 above)
Comment (calbach) [reply api/src/main/resources/workbench.yaml]: I'm not entirely following - is this at all related to this field here? https://github.com/all-of-us/workbench/blob/f8eb2419cb5d3a97c6d9db630e56775689dfaee3/api/src/main/resources/client_api.yaml#L375

Are tables containing person_ids the only tables relevant to the cohort builder?

File: api/src/main/java/org/pmiops/workbench/db/model/Cohort.java [hunk 14]
```diff
@@ -142,4 +142,18 @@ public void setCohortReviews(Set<CohortReview> cohortReviews) {
   public void addCohortReview(CohortReview cohortReview) {
     this.cohortReviews.add(cohortReview);
   }
+
+  public Cohort makeClone() {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/model/Cohort.java]: Done.

File: api/src/main/resources/workbench.yaml (hunk 5 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Thanks for pointing that enum out! It is used in concept search -- when a user asks for say "OBSERVATION" we look up in ConceptsController.DOMAIN_MAP the set of OMOP domain IDs that it maps to, and query for just those concepts.

I switched ConceptSet to use it... with logic enforcing that certain values (GENDER, RACE, and ETHNICITY) aren't valid. (These aren't like the domains that correspond to separate OMOP tables, and we don't think we need to support concept sets for them.) I like reusing the enum better than maintaining two enums I think.

I also added validation logic to ensure that concept IDs being added must have an OMOP domain ID that is mapped to by the domain on the concept set.

File: api/src/main/resources/workbench.yaml [hunk 15]
```diff
@@ -1299,6 +1421,80 @@ definitions:
...
+  UpdateConceptSetRequest:
+    type: object
+    required:
+      - etag
+    properties:
+      etag:
+        type: string
+        description: >
+          Etag of the concept set being modified.
+          Validates that the workspace (and its user roles) has not been modified since this
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Done.

File: api/src/main/resources/workbench.yaml [hunk 16]
```diff
@@ -1299,6 +1406,85 @@ definitions:
...
+      creationTime:
+        type: integer
+        format: int64
+        description: Milliseconds since the UNIX epoch.
+      lastModifiedTime:
+        type: integer
+        format: int64
+        description: Milliseconds since the UNIX epoch.
+      concepts:
+        description: "Concepts in the concept set, in alphabetical order."
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Updated.

File: api/src/main/resources/workbench.yaml [hunk 17]
```diff
@@ -1299,6 +1406,85 @@ definitions:
...
+  UpdateConceptSetRequest:
+    type: object
+    required:
+      - etag
+    properties:
+      etag:
+        type: string
+        description: >
+          Etag of the concept set being modified.
+          Validates that the concept set (and its user roles) has not been modified since this
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Done.

File: api/src/main/java/org/pmiops/workbench/db/model/ConceptSet.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+  public ConceptSet() {}
+
+  public ConceptSet(ConceptSet cs) {
+    setDescription(cs.getDescription());
+    setName(cs.getName());
+    setDomain(cs.getDomain());
+    setCreator(cs.getCreator());
+    setWorkspaceId(cs.getWorkspaceId());
+    // Should we actually copy these? (Same question for cohorts...)
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/model/ConceptSet.java]: OK, I missed that. Updated the logic to continue doing this (but now in ConceptService and ConceptSetService methods.)

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+    int version = Etags.toVersion(request.getEtag());
+    if (dbConceptSet.getVersion() != version) {
+      throw new ConflictException("Attempted to modify outdated concept set version");
+    }
+
+    if (request.getAddedIds() != null) {
+      final Domain domainEnum = dbConceptSet.getDomainEnum();
+      Iterable<org.pmiops.workbench.cdr.model.Concept> concepts = conceptDao.findAll(request.getAddedIds());
+      List<org.pmiops.workbench.cdr.model.Concept> mismatchedConcepts =
+          ImmutableList.copyOf(concepts).stream().filter(new Predicate<org.pmiops.workbench.cdr.model.Concept>() {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Done.

File: api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+        @Override
+        public boolean test(org.pmiops.workbench.cdr.model.Concept concept) {
+          Collection<Domain> domain = ConceptsController.DOMAIN_MAP.inverse().get(concept.getDomainId());
+          return domain == null || !domain.contains(domainEnum);
+        }
+      }).collect(Collectors.toList());
+      if (!mismatchedConcepts.isEmpty()) {
+        String mismatchedConceptIds = Joiner.on(", ").join(mismatchedConcepts.stream()
+            .map(org.pmiops.workbench.cdr.model.Concept::getConceptId).collect(Collectors.toList()),
+            " ,");
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java]: Oops, removed.

**Comment by calbach** [comment]
Also RE: domainId - I'm clearly out of my depth when it comes to concepts so I don't have any intelligent suggestions on this currently (without more discussion). Doing that rename in a follow-up PR (if at all), sounds ideal though.

**Comment by danqrodney** [comment]
OK, tests are updated and passing now. PTAL @calbach .

File: api/src/test/resources/application.properties [hunk 21]
```diff
@@ -3,4 +3,10 @@ spring.datasource.url=jdbc:h2:mem:db;DB_CLOSE_DELAY=-1;INIT=CREATE SCHEMA IF NOT
 spring.datasource.username=sa
 spring.datasource.password=sa
 spring.jpa.hibernate.ddl-auto=create-drop
-spring.jpa.properties.hibernate.dialect=org.pmiops.workbench.cdr.CommonTestDialect
\ No newline at end of file
+spring.jpa.properties.hibernate.dialect=org.pmiops.workbench.cdr.CommonTestDialect
+
+logging.level.org.hibernate.SQL=DEBUG
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/resources/application.properties]: Nope, removed.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,482 @@
...
+    conceptSetsController.updateConceptSetConcepts(WORKSPACE_NAMESPACE, WORKSPACE_NAME,
+        conceptSet.getId(), addConceptsRequest(conceptSet.getEtag(),
+            CLIENT_CONCEPT_1.getConceptId(),
+            CLIENT_CONCEPT_2.getConceptId()));
+  }
+
+  @Test
+  public void testDeleteConceptSet() {
+    saveConcepts();
+    ConceptSet conceptSet = makeConceptSet1();
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: Done.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,482 @@
...
+    workspace.setCdrVersionId(String.valueOf(cdrVersion.getCdrVersionId()));
+
+    Workspace workspace2 = new Workspace();
+    workspace2.setName(WORKSPACE_NAME_2);
+    workspace2.setNamespace(WORKSPACE_NAMESPACE);
+    workspace2.setDataAccessLevel(DataAccessLevel.PROTECTED);
+    workspace2.setResearchPurpose(new ResearchPurpose());
+    workspace2.setCdrVersionId(String.valueOf(cdrVersion.getCdrVersionId()));
+
+    WorkspacesController workspacesController = new WorkspacesController(workspaceService,
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: Done.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java [hunk 24]
```diff
@@ -0,0 +1,482 @@
...
+  public void testUpdateConceptSet() {
+    ConceptSet conceptSet = makeConceptSet1();
+    conceptSet.setDescription("new description");
+    conceptSet.setName("new name");
+    Instant newInstant = NOW.plusMillis(1);
+    CLOCK.setInstant(newInstant);
+    ConceptSet updatedConceptSet =
+        conceptSetsController.updateConceptSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, conceptSet.getId(),
+            conceptSet).getBody();
+    assertThat(updatedConceptSet.getCreator()).isEqualTo(USER_EMAIL);
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: I am on the fence about this.... it makes for a shorter test, but slightly harder to tell what's wrong when one field differs. When there's a lot of objects being compared, I usually use expected objects for comparison, but when it's just one I often compare individual fields like this.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java [hunk 25]
```diff
@@ -0,0 +1,482 @@
...
+    conceptSet.setId(1L);
+    conceptSet.setEtag(Etags.fromVersion(1));
+
+    conceptSetsController.updateConceptSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, 1L,
+        conceptSet);
+  }
+
+  @Test
+  public void testCreateAndGetConceptSet() {
+    ConceptSet conceptSet = makeConceptSet1();
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: So I could do that, but makeConceptSet1() gets used in a bunch of other places, too... would you copy its contents into this new method?

This didn't seem *that* confusing. :)

File: api/src/main/java/org/pmiops/workbench/db/dao/ConceptSetService.java [hunk 26]
```diff
@@ -10,6 +11,8 @@
 @Service
 public class ConceptSetService {
 
+  private static final Logger logger = Logger.getLogger("ConceptSetService");
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/dao/ConceptSetService.java]: Removed.

File: api/src/main/java/org/pmiops/workbench/db/dao/ConceptSetService.java [hunk 27]
```diff
@@ -21,8 +24,16 @@ public ConceptSet cloneConceptSetAndConceptIds(ConceptSet conceptSet, Workspace
     c.setCreator(targetWorkspace.getCreator());
     c.setLastModifiedTime(targetWorkspace.getLastModifiedTime());
     c.setCreationTime(targetWorkspace.getCreationTime());
+    c.setVersion(1);
     ConceptSet saved = conceptSetDao.save(c);
     conceptSetDao.bulkCopyConceptIds(conceptSet.getConceptSetId(), saved.getConceptSetId());
+    logger.info("FOOOOOOOO! = " + conceptSet.getConceptSetId() + ", " + saved.getConceptSetId());
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/dao/ConceptSetService.java]: Removed.

File: api/src/main/java/org/pmiops/workbench/db/model/ConceptSet.java [hunk 28]
```diff
@@ -139,7 +138,7 @@ public void setLastModifiedTime(Timestamp lastModifiedTime) {
     this.lastModifiedTime = lastModifiedTime;
   }
 
-  @ElementCollection(fetch = FetchType.LAZY)
+  @ElementCollection(fetch = FetchType.EAGER)
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/model/ConceptSet.java]: So this FetchType thing only takes effect when fetching individual entities with children, not lists of them.

If I specify LAZY here, I'll need to add another method that eagerly fetches the concepts under the concept set for the get() method, or fetches it in a second call in a @Transactional block... that didn't seem necessary since we are going to want these concepts when fetching individual concept sets generally.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: It's not *that* confusing, it just feels like a poor separation.

I'm suggesting: split this into `testCreateConceptSet()` and `testGetConceptSet()`. Move these assertions to the `create` test, and inline the `makeConceptSet1` method (or some variation) there. Continue using the helper from `getConceptSet()` here and elsewhere (but remove all these now duplicated assertions). Other methods can depend on "create" working properly, since there's now independent coverage of that.

True, there could be a bug in the specific helper function, but that's pretty low value coverage IMO.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: Sounds good, I don't have a strong opinion

File: api/src/main/java/org/pmiops/workbench/api/CohortsController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/CohortsController.java]: Done.

File: api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java]: Done.
